tensorflow/lite/micro/kernels/if.cc

```diff
@@ -111,32 +197,118 @@ TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+    int active_branch_subgraph_index) {
+  return CopyInputBetweenIfAndSubgraph(
+      context, node, graph_info, active_branch_subgraph_index,
+      CopyInputDirection::kFromIntermediateToSubgraph);
+}
+
+TfLiteStatus CopyFromIntermediateToIfOutput(TfLiteContext* context,
+                                            TfLiteNode* node,
+                                            MicroGraph& graph_info,
+                                            int active_branch_subgraph_index) {
+  return CopyOutputBetweenIfAndSubgraph(
+      context, node, graph_info, active_branch_subgraph_index,
+      CopyOutputDirection::kFromIntermediateToIf);
+}
+
+TfLiteStatus CopyFromSubgraphOutputToIntermediate(
+    TfLiteContext* context, TfLiteNode* node, MicroGraph& graph_info,
+    int active_branch_subgraph_index) {
+  return CopyOutputBetweenIfAndSubgraph(
+      context, node, graph_info, active_branch_subgraph_index,
+      CopyOutputDirection::kFromSubgraphToIntermediate);
+}
+
+TfLiteStatus Eval(TfLiteContext* context, TfLiteNode* node) {
+  const OpData* op_data = reinterpret_cast<OpData*>(node->user_data);
+
+  const TfLiteTensor* cond;
+  TF_LITE_ENSURE_OK(context, GetInputSafe(context, node, 0, &cond));
+  bool cond_value = cond->data.b[0];
+
+  tflite::MicroContext* micro_context = tflite::GetMicroContext(context);
+  MicroGraph& graph_info = micro_context->graph();
+
+  // Currently we copy the input / output between the subgraphs.
+  int active_branch_subgraph_index =
+      cond_value ? op_data->then_subgraph_index : op_data->else_subgraph_index;
+
+  // The input tensors of the IF subgraph and active branch subgraphs are
+  // planned separately and can overlap. So copying one tensor from the IF
+  // subgraph to one other tensor of the active subgraph can unintentionally
+  // overwrite another tensor of the IF subgraph. So we need to copy all input
+  // tensor of the IF subgraph to an intermediate buffer and then copied to
+  // input tensors of the active branch subgraph.
+  CopyFromIfInputToIntermediate(context, node, graph_info,
+                                active_branch_subgraph_index);
+  CopyFromIntermediateToSubgraphInput(context, node, graph_info,
+                                      active_branch_subgraph_index);
 
   TF_LITE_ENSURE_OK(context,
                     graph_info.InvokeSubgraph(active_branch_subgraph_index));
 
-  for (size_t i = 0;
-       i < graph_info.NumSubgraphOutputs(active_branch_subgraph_index); ++i) {
-    const TfLiteEvalTensor* output =
-        tflite::micro::GetEvalOutput(context, node, i);
-
-    TfLiteEvalTensor* subgraph_output =
-        graph_info.GetSubgraphOutput(active_branch_subgraph_index, i);
+  // The input tensors of the IF subgraph may already get overwritten; these
+  // tensors may still be used by other OPs and we need to restore it first from
+  // intermediate buffers.
+  CopyFromIntermediateToIfInput(context, node, graph_info,
```

Comment:
Writing to the if op's input tensors can overwrite the active subgraph's output tensors before we get a chance to copy them to the intermediate (line 307).

Reply:
I see that happen with the model that I'm testing now.

Reply:
+1 - I think that in order to avoid this issue, we need to save the intermediate inputs and have a separate intermediate output buffer.

Before invoking the subgraph, the following can happen:
1. Copy If inputs to input intermediate
2. Copy input intermediate to subgraph inputs

After invoking the subgraph the following can happen in this order:
1. Copy subgraph outputs to output intermediate
2. Copy input intermediate back to if op inputs
3. Copy output intermediate to if op outputs

Wdyt?

Reply:
Thanks Shlomy and Nat! Yes, I think you folks. It looks like we need two intermediate buffer that is NON-overlapping: one for IF input and one for Subgraph Output). I will work on this change.